clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/CompletableEventReaper.java

```diff
@@ -0,0 +1,150 @@
+ */
+public class CompletableEventReaper<T extends CompletableEvent<?>> {
+
+    private final Logger log;
+
+    /**
+     * List of tracked events that we are candidates to expire or cancel when reviewed.
+     */
+    private final List<T> tracked;
+
+    public CompletableEventReaper(LogContext logContext) {
+        this.log = logContext.logger(CompletableEventReaper.class);
+        this.tracked = new ArrayList<>();
+    }
+
+    /**
+     * Adds a new {@link CompletableEvent event} to track for later completion/expiration.
+     *
+     * @param event Event to track
+     */
+    public void add(T event) {
+        tracked.add(Objects.requireNonNull(event, "Event to track must be non-null"));
+    }
+
+    /**
+     * This method "completes" any {@link CompletableEvent}s that have either expired or completed normally. So this
+     * is a two-step process:
+     *
+     * <ol>
+     *     <li>
+     *         For each tracked event which has exceeded its {@link CompletableEvent#deadlineMs() deadline}, an
+     *         instance of {@link TimeoutException} is created and passed to
+     *         {@link CompletableFuture#completeExceptionally(Throwable)}.
+     *     </li>
+     *     <li>
+     *         For each tracked event of which its {@link CompletableEvent#future() future} is already in the
+     *         {@link CompletableFuture#isDone() done} state, it will be removed from the list of tracked events.
+     *     </li>
+     * </ol>
+     *
+     * <p/>
+     *
+     * This method should be called at regular intervals, based upon the needs of the resource that owns the reaper.
+     *
+     * @param currentTimeMs <em>Current</em> time with which to compare against the
+     *                      <em>{@link CompletableEvent#deadlineMs() expiration time}</em>
+     */
+    public void reapExpiredAndCompleted(long currentTimeMs) {
+        log.trace("Reaping expired events");
+
+        Consumer<CompletableEvent<?>> timeoutEvent = e -> {
+            TimeoutException error = new TimeoutException(String.format("%s could not be completed within its timeout", e.getClass().getSimpleName()));
+            long pastDueMs = currentTimeMs - e.deadlineMs();
+            log.debug("Completing event {} exceptionally since it expired {} ms ago", e, pastDueMs);
+            CompletableFuture<?> f = e.future();
+            f.completeExceptionally(error);
+        };
+
+        // First, complete (exceptionally) any events that have passed their deadline AND aren't already complete.
+        tracked.stream()
+            .filter(e -> !e.future().isDone())
+            .filter(e -> currentTimeMs > e.deadlineMs())
+            .forEach(timeoutEvent);
+        // Second, remove any events that are already complete, just to make sure we don't hold references. This will
+        // include any events that finished successfully as well as any events we just completed exceptionally above.
+        tracked.removeIf(e -> e.future().isDone());
```

Comment:
Couldn't we make that the same add operation removes the event whenComplete? Seems tighter that the same operation that adds the event ensures that it's removed (if completes), and then it's simpler here, when we only need to care about maintaining the uncompleted (which seems like the core responsibility of the reaper).

Reply:
@lianetm Do you propose to remove completed events in `CompletableEventReaper#add`?
If yes, to avoid the `ConcurrentModifictionException` @kirktrue, you could iterate over `tracked` with an iterator and remove completed events through the iterator:
```java
final Iterator<T> iterator = tracked.iterator();
while (iterator.hasNext()) {
    final T event = iterator.next();
    if (event.future().isDone()) {
        iterator.remove();
    }
}
```

Reply:
On a second thought, what is the issue with moving 
```java
tracked.removeIf(e -> e.future().isDone());
```
to `add()`?

I guess, I am misunderstanding your comments.

Reply:
Keep in mind that `add()` is called _for each event_ in the queue. If we remove completed events in `add()`, we're updating `tracked` each time, too.

In the current approach, `tracked` is only updated once inside `reapExpiredAndCompleted()`.

I think I'm missing the benefit of moving it to `add()` 🤔

Reply:
the benefit I see is self-cleaning which reduces the scope/responsibilities of `tracked`. We don't really care about events that are created and complete successfully (so nice that they are just temporarily tracked and automatically disappear when they complete). Then `reapExpiredAndCompleted` it's simply about `reapExpired`

Reply:
just a suggestion but I would leave it to you as they may be impl details I'm missing ;)

Reply:
@lianetm & @cadonna—please let me know if the following makes sense. I am trying to convince myself of this design as much as anyone else 😄...

-------

The `ConsumerNetworkThread.run()` method sits in a tight loop that calls `runOnce()` on each pass. The ordering of operations inside `runOnce()` is as follows:

1. Call `processApplicationEvents()`, for each event... 
  a. Call `CompletableEventReaper.add()` to add the event to `tracked` list
  b. Call `ApplicationEventProcessor.process()` to call relevant request manager APIs based on the event. Many of the calls to the request manager APIs will create and return `CompletableFuture`s
2. Call `RequestManager.poll()` for each request manager
3. Call `NetworkClientDelegate.add()` for each `UnsentRequest` returned from step 2
4. Call `NetworkClientDelegate.poll()` to process all the requests added in step 3. This will invoke handlers for received network responses, which will call `complete()`/`completeExceptionally()` to be called on any `CompletableFuture`s created in step 1b
5. Call `reapExpiredApplicationEvents()` (which calls `CompletableEventReaper.reap()`) to remove expired events and any `CompletableFuture`s completed in step 4

The design of `ConsumerNetworkThread` is such that an event's `Future` will _only_ `complete()` (success or failure) in step 4. Any events eligible to be removed from `tracked` in step 1 would have already been removed in the previous loop's step 5. So it makes the most sense to me to leave the removal in `reap()` vs. `add()`.

Reply:
Yes, the flow makes sense to me. Also fine with me to keep the reap handling completed and expired.